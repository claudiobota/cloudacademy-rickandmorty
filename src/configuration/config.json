{
	"api": {
		"url": "https://rickandmortyapi.com/api",
		"paths": {
			"characters": "/character",
			"locations": "/location",
			"episodes": "/episode"
		},
		"page_size": 20
	}
}
