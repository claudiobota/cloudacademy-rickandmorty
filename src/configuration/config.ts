import configData from './config.json';
import configDevelopmentData from './config.development.json';
import configTestData from './config.test.json';
import configStagingData from './config.staging.json';
import configProductionData from './config.production.json';
import {IConfig} from '../types/interfaces';
import {Environment} from '../types/enums';

let config: IConfig = {
  ...configData
};

const nodeEnv: string = process.env.NODE_ENV || 'development';

switch (nodeEnv) {
  case Environment.DEVELOPMENT:
    config = {
      ...config,
      ...configDevelopmentData
    };
    break;

  case Environment.TEST:
    config = {
      ...config,
      ...configTestData
    };
    break;

  case Environment.STAGING:
    config = {
      ...config,
      ...configStagingData
    };
    break;

  case Environment.PRODUCTION:
    config = {
      ...config,
      ...configProductionData
    };
    break;

  default:
}

export default config;
