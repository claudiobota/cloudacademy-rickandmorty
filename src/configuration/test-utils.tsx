import React from 'react';
import {render as rtlRender} from '@testing-library/react';
import {Provider} from 'react-redux';
import {configureAppStore} from '../redux/store';

function render(
  ui: React.ReactElement,
  {
    initialState,
    store = configureAppStore(initialState),
    ...renderOptions
  }: any = {}
) {
  function Wrapper({ children }: { children: React.ReactElement }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render }
