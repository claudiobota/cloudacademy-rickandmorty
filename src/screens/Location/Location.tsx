import React, {useEffect, useState} from 'react';
import {Link, useParams } from 'react-router-dom';
import {ICharacter, ILocation} from '../../types/interfaces';
import api from '../../services/api';
import {ProgressSpinner} from 'primereact/progressspinner';
import {Card} from 'primereact/card';
import {Button} from 'primereact/button';
import CharactersListing from '../../components/CharactersListing/CharactersListing';
import config from '../../configuration/config';

export default function Location() {
  const { id } = useParams<{id: string;}>();
  const [location, setLocation] = useState<ILocation | undefined>(undefined);
  const [characters, setCharacters] = useState<ICharacter[]>([]);

  useEffect(() => {
    api.retrieveLocation(Number(id)).then(location => {
      setLocation(location);
    }).catch();
  }, [id]);

  useEffect(() => {
    if (location) {
      api.retrieveCharactersByIds(getLocationResidentsIds()).then(characters => {
        setCharacters(characters || []);
      }).catch();
    }
  }, [location]);

  const getLocationSubtitle = () => {
    return (
      <div className="p-grid">
        <div className="p-sm-4 p-d-flex p-flex-row p-align-center">
          <span>Dimension</span>
          <strong className="p-d-block p-ml-3">{location?.dimension}</strong>
        </div>
        <div className="p-sm-4 p-d-flex p-flex-row p-align-center">
          <span>Type</span>
          <strong className="p-d-block p-ml-3">{location?.type}</strong>
        </div>
        <div className="p-sm-4 p-d-flex p-flex-row p-align-center">
          <span>Total Residents</span>
          <strong className="p-d-block p-ml-3">{location?.residents.length}</strong>
        </div>
      </div>
    )
  };

  const getLocationResidentsIds = () => {
    const residentIds = location?.residents.map(residentUrl => Number(residentUrl.replace(`${config.api.url}${config.api.paths.characters}/`, '')));
    return residentIds || [];
  }

  let content;

  if (!location) {
    content = (
      <Card className="p-text-center" style={{ width: '100%' }}>
        <ProgressSpinner style={{width: '50px', height: '50px'}} fill="#FFFFFF" />
      </Card>
    );
  } else {
    content = (
      <div className="wrapper">
        <Card title={location.name} subTitle={getLocationSubtitle()}>
          <Link to="/">
            <Button label="Return home" icon="pi pi-arrow-left" className="p-button-secondary p-button-outlined" />
          </Link>
        </Card>
        <div className="p-mt-5">
          <h1 style={{ color: '#FFFFFF' }}>Residents</h1>
          <CharactersListing characters={characters} />
        </div>
      </div>
    );
  }

  return (
    <div className="container p-d-flex p-flex-column">
      {content}
    </div>
  )
}
