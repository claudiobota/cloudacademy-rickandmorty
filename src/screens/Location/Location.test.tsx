import { createMemoryHistory } from 'history';
import React from 'react';
import {render} from '../../configuration/test-utils';
import Location from './Location';
import {Route, Router} from 'react-router-dom';

describe('Location', () => {
  test('screen renders and matches snapshot', () => {
    const history = createMemoryHistory();
    const url = '/location/1';
    history.push(url);
    const { asFragment } = render(
      <Router history={history}>
        <Route path="/location/:id">
          <Location />
        </Route>
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
