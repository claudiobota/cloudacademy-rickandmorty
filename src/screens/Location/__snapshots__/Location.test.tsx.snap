// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location screen renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container p-d-flex p-flex-column"
  >
    <div
      class="p-card p-component p-text-center"
      style="width: 100%;"
    >
      <div
        class="p-card-body"
      >
        <div
          class="p-card-content"
        >
          <div
            aria-busy="true"
            class="p-progress-spinner"
            role="alert"
            style="width: 50px; height: 50px;"
          >
            <svg
              class="p-progress-spinner-svg"
              style="animation-duration: 2s;"
              viewBox="25 25 50 50"
            >
              <circle
                class="p-progress-spinner-circle"
                cx="50"
                cy="50"
                fill="#FFFFFF"
                r="20"
                stroke-miterlimit="10"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
